FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RecipientsViewModel.kt

```diff
@@ -0,0 +1,391 @@
+            val emailLowerCase = email.lowercase(Locale.getDefault())
+            var cachedRecipientWithPubKeys = getCachedRecipientWithPubKeys(emailLowerCase)
+
+            if (cachedRecipientWithPubKeys == null) {
+              roomDatabase.recipientDao().insertSuspend(RecipientEntity(email = emailLowerCase))
+              cachedRecipientWithPubKeys =
+                roomDatabase.recipientDao().getRecipientWithPubKeysByEmailSuspend(emailLowerCase)
+            } else {
+              for (publicKeyEntity in cachedRecipientWithPubKeys.publicKeys) {
+                try {
+                  val result = PgpKey.parseKeys(publicKeyEntity.publicKey).pgpKeyDetailsList
+                  publicKeyEntity.pgpKeyDetails = result.firstOrNull()
+                } catch (e: Exception) {
+                  e.printStackTrace()
+                  publicKeyEntity.isNotUsable = true
+                }
+              }
+            }
+
+            try {
+              getPgpInfoFromServer(email = emailLowerCase)?.let { pgpKeyDetailsList ->
+                cachedRecipientWithPubKeys?.let { recipientWithPubKeys ->
+                  updateCachedInfoWithPubKeysFromLookUp(
+                    recipientWithPubKeys,
+                    pgpKeyDetailsList
+                  )
+                }
+              }
+              cachedRecipientWithPubKeys = getCachedRecipientWithPubKeys(emailLowerCase)
+              cachedRecipientWithPubKeys?.let { recipients.add(it) }
+            } catch (e: Exception) {
+              e.printStackTrace()
+              ExceptionUtil.handleError(e)
+            }
+          }
+        }
+        setResultForRemoteContactsLiveData(type, Result.success(recipients))
+      } catch (e: Exception) {
+        e.printStackTrace()
+        ExceptionUtil.handleError(e)
+        setResultForRemoteContactsLiveData(type, Result.exception(e))
+      }
+    }
+  }
+
+  private suspend fun getCachedRecipientWithPubKeys(emailLowerCase: String): RecipientWithPubKeys? =
+    withContext(Dispatchers.IO) {
+      val cachedRecipientWithPubKeys = roomDatabase.recipientDao()
+        .getRecipientWithPubKeysByEmailSuspend(emailLowerCase) ?: return@withContext null
+
+      for (publicKeyEntity in cachedRecipientWithPubKeys.publicKeys) {
+        try {
+          val result = PgpKey.parseKeys(publicKeyEntity.publicKey).pgpKeyDetailsList
+          publicKeyEntity.pgpKeyDetails = result.firstOrNull()
+        } catch (e: Exception) {
+          e.printStackTrace()
+          publicKeyEntity.isNotUsable = true
+        }
+      }
+      return@withContext cachedRecipientWithPubKeys
+    }
+
+  private suspend fun updateCachedInfoWithPubKeysFromLookUp(
+    cachedRecipientEntity: RecipientWithPubKeys, list: List<PgpKeyDetails>
+  ) = withContext(Dispatchers.IO) {
+    val email = cachedRecipientEntity.recipient.email
+    val mapOfUniquePubKeysFromLookUp = mutableMapOf<String, PgpKeyDetails>()
+
+    for (pgpKeyDetails in list) {
```

Comment:
This whole `for` loop should be put inside it's own private method that should be called `deduplicateFetchedPubKeysByFingerprint` that returns a list. So you hide the map inside of the method. It took me a whole to figure out what is going on here, giving it a name like this will make it much clearer.